sdks/python/apache_beam/metrics/cells.py

```diff
@@ -0,0 +1,200 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import threading
+
+from apache_beam.metrics.base import Counter
+from apache_beam.metrics.base import Distribution
+
+
+class DirtyState(object):
+  """ Keeps track of a cell's commit status
+
+  It's thread-safe.
+  """
+  DIRTY = 0
+  CLEAN = 1
+  COMMITTING = 2
+
+  def __init__(self):
+    self._lock = threading.Lock()
+    self._state = DirtyState.DIRTY
+
+  @property
+  def state(self):
+    return self._state
+
+  def modified(self):
+    with self._lock:
+      self._state = DirtyState.DIRTY
+
+  def after_commit(self):
+    with self._lock:
+      if self._state == DirtyState.COMMITTING:
+        self._state = DirtyState.CLEAN
+
+  def before_commit(self):
+    with self._lock:
+      if self._state == DirtyState.CLEAN:
+        return False
+      else:
+        self._state = DirtyState.COMMITTING
+        return True
+
+
+class MetricCell(object):
+  """ Base class of a Cell that tracks the state of a metric.
+
+  It's thread safe.
+  """
+  def __init__(self):
+    self.dirty = DirtyState()
+    self._lock = threading.Lock()
+
+  def get_cumulative(self):
+    raise NotImplementedError
+
+
+class CounterCell(Counter, MetricCell):
+  """ Tracks the state of a counter metric.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(CounterCell, self).__init__(*args)
+    self.value = 0
+
+  def combine(self, other):
+    result = CounterCell()
+    result.inc(self.value + other.value)
+    return result
+
+  def inc(self, n=1):
+    with self._lock:
+      self.value += n
+      self.dirty.modified()
+
+  def get_cumulative(self):
+    return self.value
+
+
+class DistributionCell(Distribution, MetricCell):
+  """ Tracks the state of a distribution metric.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(DistributionCell, self).__init__(*args)
+    self.data = DistributionData(0, 0, None, None)
+
+  def combine(self, other):
+    result = DistributionCell()
+    result.data = self.data.combine(other.data)
+    return result
+
+  def update(self, value):
+    with self._lock:
+      self.dirty.modified()
```

Comment:
May be worth having the dirty-bit also have the lock. That way we don't need to get two locks to do the dirty-bit setting within the update method. The downside is that getting metrics then has the same lock...

Alternatively, is there an atomic long or anything like that that could be done via compare-and-swap?